src/gpuarray/array.h

```diff
@@ -607,6 +607,78 @@ GPUARRAY_PUBLIC void GpuArray_fprintf(FILE *fd, const GpuArray *a);
 
 GPUARRAY_PUBLIC int GpuArray_fdump(FILE *fd, const GpuArray *a);
 
+/************************************************************************************
+*                         multi-gpu collectives interface                          *
+************************************************************************************/
+
+/**
+ * \brief TODO
+ * \param src [const GpuArray*] TODO
+ * \param dest [GpuArray*] TODO
+ * \param count [int] TODO
+ * \param typecode [int] TODO
+ * \param opcode [int] TODO
+ * \param root [int] TODO
+ * \param comm [gpucomm*] TODO
+ * \return int TODO
+ */
+GPUARRAY_PUBLIC int GpuArray_reduce(const GpuArray* src, GpuArray* dest,
+                                    int count, int typecode, int opcode,
+                                    int root, gpucomm* comm);
```

Comment:
I agree that the reduce operations would need to take a gpucomm parameter to know with what to reduce.  What I'm proposing is that we only build the low-level interface that takes gpucomm and gpudata for now, and think about a high-level one later.

I agree with you that passing an index or something like that is a bad idea.